---
title: "maize GRN figures and tables"
output: 
  pdf_document:
    citation_package: natbib
    keep_tex: false
    fig_caption: true
    latex_engine: pdflatex
    template: rmd-latex-ms.tex
# author:
# - name: Peng Zhou
#  affiliation: University of Minnesota
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathpazo
fontsize: 11pt
# spacing: double
# biblio-style: apsr
header-includes:
- \usepackage{caption}
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage[table]{xcolor}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \captionsetup[figure]{labelformat=empty}
- \captionsetup[table]{labelformat=empty}
- \AtBeginDocument{\let\maketitle\relax}
---
\pagenumbering{gobble}


```{r, message=FALSE, echo=FALSE}
require(tidyverse)
require(knitr)
require(kableExtra)
options(knitr.kable.NA = '')
dirw = '~/projects/maize.grn/data' 
ft = file.path(dirw, "10.dataset.tsv")
tt = read_tsv(ft)
kable(tt, format = 'latex', longtable = T, booktabs = T,
  format.args = list(big.mark = ","), 
  caption = "Table X. GRNs built in this study.") %>%
  column_spec(1, bold=F) %>%
  kable_styling("striped", full_width = F)
```


\pagebreak
![Fig 1]($grn/data/13_eval/10.dev.atlas.pdf){width=100%}

Area under receiver operating curves (AUROC) and area under precision-recall curve (AUPR) for GRNs built using different input datasets evaluated using experimentally (Chip-seq, mutant & wildtype RNA-Seq) determined transcription factor (TF) targets.

\pagebreak
![Fig 2]($grn/data/13_eval/05.auc.pdf){width=100%}

Barplot showing AUROC and AUPRs for lin2017 tissue-specific and pooled GRNs. 

\pagebreak
![Fig 3a]($grn/data/13_eval/12.briggs.1e4.pdf){width=100%}

Enrichment of DE genes (using Briggs dataset) in predicted TF targets made by different GRNs. Only the top 10,000 strongest edges were taken from each GRN.

\pagebreak
![Fig 3b]($grn/data/13_eval/12.briggs.1e5.pdf){width=100%}

Enrichment of DE genes (using Briggs dataset) in predicted TF targets made by different GRNs. Only the top 100,000 strongest edges were taken from each GRN.

\pagebreak
![Fig 3c]($grn/data/13_eval/12.briggs.1e6.pdf){width=100%}

Enrichment of DE genes (using Briggs dataset) in predicted TF targets made by different GRNs. Only the top 1,000,000 strongest edges were taken from each GRN.

\pagebreak
![Fig 4a]($grn/data/13_eval/15.biomap.all.pdf){width=100%}

Evaluation of GRN predictions using the bioMAP data containing ~200 genotypes (34 inbreds + 200 hybrids) in five tissues.  The (Pearson) correlations between regulators and targets (as predicted by each GRN) were evaluated using the bioMAP expression matrix in each tissue, and shown as a boxplot for the top 10,000, 100,000 or 1,000,000 strongest edges in each network.

\pagebreak
![Fig 4b]($grn/data/13_eval/15.biomap.pdf){width=100%}

Evaluation of GRN predictions using the bioMAP data containing ~200 genotypes (34 inbreds + 200 hybrids) in five tissues.  The (Pearson) correlations between regulators and targets (as predicted by each GRN) were evaluated using the bioMAP expression matrix in each tissue, and shown as a boxplot for the top 10,000 strongest edges in each network.
