---
title: "maize GRN figures and tables"
output:
  pdf_document:
    citation_package: natbib
    keep_tex: false
    fig_caption: true
    latex_engine: pdflatex
    template: rmd-latex-ms.tex
# author:
# - name: Peng Zhou
#  affiliation: University of Minnesota
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathptmx
fontsize: 11pt
# spacing: double
# biblio-style: apsr
header-includes:
- \usepackage{caption}
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage[table]{xcolor}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \captionsetup[figure]{labelformat=empty}
- \captionsetup[table]{labelformat=empty}
- \AtBeginDocument{\let\maketitle\relax}
- \usepackage{lscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
---
\pagenumbering{gobble}

```{r, message=FALSE, echo=FALSE}
dirw = '/home/springer/zhoux379/projects/grn/data/12_tables'
```

![]($grn/data/13_eval/02.gopt.1.hc.pdf){width=100%}

Figure S1. Hierarchical clustering of 132 GRNs (44 expression dataset * 3 methods).  Each of the three regression-based method (RF - random forest, ET - extra trees, XGB - gradient boosted) was used to build a network for each RNA-Seq dataset.  Pairwise distance between networks was determined by taking the top 100,000 TF-target predictions from each network and calculating the proportion of shared (common) predictions (using ‘dist()’ function in R with additional argument ‘method=binary’).  Hierarchical clustering was then performed based on the cross-network pairwise distance matrix using “ward.D” option.

\pagebreak
![]($grn/data/13_eval/11.tf.pdf){width=100%}

Figure S2.  Comparison of GRNs built with different methods using supports from known direct targets of published TF studies in maize.  Direct targets of 7 TFs were obtained from literature and treated as ground truth to evaluate: (A) Area under receiver-operating characteristic curve (AUROC) evaluated for each GRN (until an False Positive Rate of 0.1 is reached) and (B) Wilcox rank test using the predicted (TF-target) interaction scores between the group of true targets and non-targets. Numbers in each cell show the actual AUROC values / test P-value (-log10 transformed) with blank cells standing for “not significant” (P > 0.05). White cells stand for missing data where the TF being evaluated is not expressed in the corresponding GRN.  Each GRN was built using three regression methods: RF (random forest), ET (extra trees) and XGB (gradient boosting).

\pagebreak
![]($grn/data/13_eval/11.ko.pdf){width=100%}

Figure S3.  Comparison of GRNs built with different methods using supports from TF knockout mutant RNA-Seq data.  Differentially expressed genes between each one of the 21 TF knockout mutants and the wildtype were identified using DESeq2 (p-value < 0.01) and treated as ground truth to evaluate: (A) Area under receiver-operating characteristic curve (AUROC) evaluated for each GRN (until an False Positive Rate of 0.1 is reached) and (B) Wilcox rank test using the predicted (TF-target) interaction scores between the group of true targets and non-targets. Numbers in each cell show the actual AUROC values / test P-value (-log10 transformed) with blank cells standing for “not significant” (P > 0.05). White cells stand for missing data where the TF being evaluated is not expressed in the corresponding GRN.  Each GRN was built using three regression methods: RF (random forest), ET (extra trees) and XGB (gradient boosting).

\pagebreak
![]($grn/data/13_eval/15.go.heat.pdf){width=80%}

Figure S4.  Comparison of GRNs built with different methods according to enrichment of functional annotations (Gene Ontology, CornCyc, etc).  For each network only the top 100,000 predicted TF-target associations were taken.  Fold enrichment is calculated as the observed number of shared GO/CornCyc terms (by targets regulated by a common TF) divided by the expected number of shared annotation terms (determined by permutation). The names for each of the networks are color coded to indicate B73 developmental surveys (blue), genotype surveys (red), meta-networks (green) or previously generated GRNs (teal).  Each GRN was built using three regression methods: RF (random forest), ET (extra trees) and XGB (gradient boosting).

\pagebreak
![]($grn/data/14_eval_sum/05.auc.pdf){width=100%}

Figure S5. Area under receiver-operating characteristic curve (AUROC) evaluated for each GRN (until an False Positive Rate of 0.1 is reached) using support from (A) direct TF targets and (B) 21 maize TF knockout mutants as ground truth.  Differentially expressed genes between each TF knockout mutant and the wildtype were identified using DESeq2 (p-value < 0.01) and treated as ground truth to evaluate GRN performance.  Numbers in each cell show the actual AUROC values with white cells indicating missing data (the TF being filtered from the expression matrix due to invariable expression).

\pagebreak
![]($grn/data/14_eval_sum/06.tfbs.pdf){width=80%}

Figure S6.  Evaluation (AUROC and Wilcox P-value) of constructed GRNs using four sets of predicted TF-target interactions based on TF-binding site motif, conserved element of TFBS motif or FunTFBS.  There are 202,670 “motif”-based predictions, 37,699 predictions based on motif and cross-species conservation, 7,459 predictions based on motif and FunTFBS as well as 4,115 predictions based on all three evidences.

\pagebreak
![]($grn/data/14_eval_sum/07.go.all.pdf){width=100%}

Figure S7.  Enrichment of co-annotated GO/CornCyc terms in co-regulated network targets.  For each network the top 1 million predicted TF-target associations were binned to 10 bins and assessed for enrichment of GO/CornCyc functional annotation.  Fold enrichment is calculated as the observed number of shared GO/CornCyc terms (by targets regulated by a common TF) divided by the expected number of shared annotation terms (determined by permutation). The names for each of the networks are color coded to indicate B73 developmental surveys (blue), genotype surveys (red), meta-networks (green) or previously generated GRNs (teal). A total of six sources of GO annotation were used: GO_HC (high quality hand-curated terms transferred from maize AGP_v3 annotation), GO_arabidopsis, GO_uniprot.plants, GO_Interproscan5, GO_pannzer and GO_argot2.5.

\pagebreak
![]($grn/data/14_eval_sum/03.tsne.pdf){width=80%}

Figure S8.  T-SNE clustering of 44 GRNs.  Top 500,000 TF-target predictions were extracted from each network to perform t-SNE clustering using parameter “perplexing=9, permutation=2000”.

\pagebreak
![]($grn/data/14_eval_sum/11.nv.1s.pdf){width=100%}

Figure S9.  TF-target validation of the combined tissue  network in all six selected natural variation datasets.  Each panel shows the proportion of differentially expressed targets regulated by TFs showing different DE levels between two genotypes in one tissue/treatment condition.  TF-target predictions were binned to 10 groups based on the interaction score in GRN.  Each TF-target pair is classified according to the DE level of the TF (“non_DE”, “DE1-2”, “DE2-4”, “DE4+” or “SPE”) in each network.  The proportion of TF-target pairs with the target also showing DE was then determined for each category.  Dashed line in each panel represents the genome-wide (background) proportion of DE genes in each tissue/treatment setting.

\pagebreak
![]($grn/data/14_eval_sum/30.eQTL.pdf){width=100%}

Figure S10. Enrichment of co-regulated targets between previously identified trans-eQTL hotspots and TF-target associations predicted by GRNs.  For each network the top 1 million TF-target predictions were binned to 10 groups based on the interaction score in GRN.  Fold enrichment is determined by the same permutation approach described in Figure 2.

\pagebreak
![]($grn/data/14_eval_sum/31.eQTL.heat.pdf){width=50%}

Figure S11. Enrichment of co-regulated targets between previously identified trans-eQTL hotspots and TF-target associations predicted by GRNs.  For each network the top 1 million predicted TF-target associations were binned to 10 bins and only the first bin (top 100k edges) were used to assess enrichment.  15 high quality networks (marked with crosses) were selected to identify the main TF regulators underlying trans-eQTL hotspots (see Methods).

\pagebreak
![]($grn/data/14_eval_sum/32.hs.pdf){width=100%}

Figure S12.  Co-localization of TFs predicted by GRNs in this study and trans-eQTL hotspots identified in previous studies that regulate the same set of targets.  Each trans-eQTL hotspot were first tested for significant overlap in targets with any TFs (hypergeometric enrichment test as implemented in the phyper() function in R).  To control for false positives, only TFs identified in at least two (out of 8 high quality networks, see Methods) that show significant co-regulation with at least one trans-eQTL hotspot (p < 0.01) were kept.  Trans-eQTL hotspots identified in previous maize assemblies were lifted over to the AGPv4 assembly coordinates.  Co-localization of a TF and a trans-eQTL is determined if the two coordinates are within 10-Mbp distance.  Color of each dot represents the number of common targets between the predicted TF regulator and the trans-eQTL hotspot.

\pagebreak
Table S1. TF knockout mutant RNA-Seq datasets used in this study.
```{r, message=FALSE, echo=FALSE}
readRDS(file.path(dirw, '01.t2.rds'))
```

\pagebreak
Table S2. Natural variation datasets used for validation in this study.
```{r, message=FALSE, echo=FALSE}
readRDS(file.path(dirw, '01.t3.rds'))
```

\pagebreak
Table S3. GRN-predicted TFs supported by trans-eQTL hotspots.
```{r, message=FALSE, echo=FALSE}
readRDS(file.path(dirw, '01.t9.rds'))
```
