---
title: "maize GRN figures and tables"
output:
  pdf_document:
    citation_package: natbib
    keep_tex: false
    fig_caption: true
    latex_engine: lualatex
# author:
# - name: Peng Zhou
#  affiliation: University of Minnesota
date: "`r format(Sys.time(), '%B %d, %Y')`"
fontsize: 10pt
geometry: "left=0.5in,right=0.5in,top=0.8in,bottom=0.5in"
# spacing: double
# biblio-style: apsr
header-includes:
- \usepackage{fontspec}
- \setmainfont[Scale=MatchUppercase]{TeX Gyre Heros}
- \usepackage{fancyhdr}
- \usepackage{caption}
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \captionsetup[figure]{labelformat=empty}
- \captionsetup[table]{labelformat=empty}
- \AtBeginDocument{\let\maketitle\relax}
- \usepackage{lscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
- \pagestyle{fancy}
- \fancyhf{} % sets both header and footer to nothing
- \renewcommand{\headrulewidth}{0pt}
- \fancyhead[L]{Supplemental Data. Zhou et al. (2020). Plant Cell 10.1105/tpc.20.00080}
---

```{r, message=FALSE, echo=FALSE}
dirw = '/home/springer/zhoux379/projects/grn/data/12_tables'
```

![]($grn/data/13_eval/15.go.heat.pdf){width=80%}

Supplemental Figure 1.  Supplemental Figure 1.  Comparison of GRNs built using different methods according to the enrichment of functional annotations (Gene Ontology, CornCyc, and so on) (Supports Figure 2).  For each network only the top 100,000 predicted TF-target associations were taken.  Fold enrichment is calculated as the observed number of shared GO/CornCyc terms (by targets regulated by a common TF) divided by the expected number of shared annotation terms (determined by permutation). The names for each of the networks are color-coded to indicate B73 developmental surveys (blue), genotype surveys (red), meta-networks (green) or previously generated GRNs (teal).  Each GRN was built using three regression methods: RF (random forest), ET (extra trees) and XGB (gradient boosting).

\pagebreak
![]($grn/data/14_eval_sum/05.tp.pdf){width=100%}

Supplemental Figure 2.  Number of true TF targets captured by the top one million predictions and the top 100K predictions in each GRN (Supports Figure 2 and 3). (A) and (C): direct targets of published TF studies derived from ChIP-Seq and mutant RNA-Seq experiments; (B) and (D) For each one of the 17 maize TFs with knockout mutant RNA-Seq data available, differentially expressed genes between mutant and wildtype were identified using DESeq2 (p-value < 0.01).  Numbers in each cell represent the actual number of true targets captured by each GRN during each evaluation, while cells were colored based on the proportion of captured true targets.  Blank (white) cells stand for missing data where the TF being evaluated is not expressed or not variable (i.e., zero variance) in the corresponding GRN.  Y-axis labels correspond to the different networks listed in Table 1.  X-axis labels (e.g., "KN1_ear (272)" or "KN1_ear [1576] [7.0%]") represent the common name for each TF, the tissue in which the TF is expressed, followed by the number of direct targets (Panels A and C) or number and proportion of differentially expressed genes in TF mutant (Panels B and D).

\pagebreak
![]($grn/data/14_eval_sum/05.cp.pdf){width=80%}

Supplemental Figure 3. Evaluation of GRNs using support from direct targets of eight known TFs (Supports Figure 2A). (A) Area under receiver-operating characteristic curve (AUROC) until an False Positive Rate of 0.1 is reached; (B) Wilcox rank test performed using the predicted (TF-target) interaction scores between the group of true targets (DEGs) and non-targets (non-DEGs); (C) Area under receiver-operating characteristic curve (AUROC) and (D) Area under precision-recall curve (AUPRC).  Light yellow cells with no numbers stand for “not significant” (P > 0.05), while blank (white) cells stand for missing data where the TF being evaluated is not expressed or not variable (i.e., zero variance) in the corresponding GRN.

\pagebreak
![]($grn/data/14_eval_sum/05.dp.pdf){width=100%}

Supplemental Figure 4. Evaluation of GRNs using support from 31 maize TF DAP-Seq datasets (Supports Figure 2). (A) Area under receiver-operating characteristic curve (AUROC) until an False Positive Rate of 0.1 is reached; (B) Wilcox rank test performed using the predicted (TF-target) interaction scores between the group of true targets (DEGs) and non-targets (non-DEGs); (C) Area under receiver-operating characteristic curve (AUROC) and (D) Area under precision-recall curve (AUPRC).  Light yellow cells with no numbers stand for “not significant” (P > 0.05), while blank (white) cells stand for missing data where the TF being evaluated is not expressed or not variable (i.e., zero variance) in the corresponding GRN.

\pagebreak
![]($grn/data/14_eval_sum/05.ko.pdf){width=100%}

Supplemental Figure 5. Evaluation of GRNs using support from 17 maize TF knockout mutant RNA-Seq datasets (Supports Figure 2). (A) Area under receiver-operating characteristic curve (AUROC) until an False Positive Rate of 0.1 is reached; (B) Wilcox rank test performed using the predicted (TF-target) interaction scores between the group of true targets (DEGs) and non-targets (non-DEGs); (C) Area under receiver-operating characteristic curve (AUROC) and (D) Area under precision-recall curve (AUPRC).  Light yellow cells with no numbers stand for “not significant” (P > 0.05), while blank (white) cells stand for missing data where the TF being evaluated is not expressed or not variable (i.e., zero variance) in the corresponding GRN.


\pagebreak
![]($grn/data/14_eval_sum/07.go.all.pdf){width=100%}

Supplemental Figure 6.  Enrichment of co-annotated GO/CornCyc terms in co-regulated network targets (Supports Figure 3A).  For each network the top 1 million predicted TF-target associations were binned to 10 bins and assessed for enrichment of GO/CornCyc functional annotation.  Fold enrichment is calculated as the observed number of shared GO/CornCyc terms (by targets regulated by a common TF) divided by the expected number of shared annotation terms (determined by permutation). The names for each of the networks are color coded to indicate B73 developmental surveys (blue), genotype surveys (red), meta-networks (green) or previously generated GRNs (teal). A total of six sources of GO annotation were used: GO_HC (high quality hand-curated terms transferred from maize AGP_v3 annotation), GO_arabidopsis, GO_uniprot.plants, GO_Interproscan5, GO_pannzer and GO_argot2.5.


\pagebreak
![]($grn/data/16_pathways/22b.pdf){width=100%}

Supplemental Figure 7.  Different GRNs capture distinct parts of documented transcriptional regulation interactions from Arabidopsis for the abscisic acid (ABA) pathway and HY5 (Elongated Hypocotyl 5) regulated pathway (Supports Figure 4B and 4C).

\pagebreak
![]($grn/data/14_eval_sum/06.tfbs.pdf){width=80%}

Supplemental Figure 8.  Evaluation (AUROC and Wilcox P-value) of constructed GRNs using four sets of predicted TF-target interactions based on TF-binding site motif, conserved element of TFBS motif, or FunTFBS (Supports Figure 2).  There are 202,670 “motif”-based predictions, 37,699 predictions based on motif and cross-species conservation, 7,459 predictions based on motif and FunTFBS as well as 4,115 predictions based on all three types of evidence.

\pagebreak
![]($grn/data/14_eval_sum/03.hc.pdf){width=100%}

Supplemental Figure 9.  Hierarchical clustering of 45 GRNs (Supports Figure 4 and 5).  Pairwise distance between networks was determined by taking the top 100,000 TF-target predictions from each network and calculating the proportion of shared (common) predictions (using ‘dist()’ function in R with additional argument ‘method=binary’).  Hierarchical clustering was then performed based on the cross-network pairwise distance matrix using “ward.D” method.

\pagebreak
![]($grn/data/14_eval_sum/03.tsne.pdf){width=80%}

Supplemental Figure 10.  T-SNE clustering of 45 GRNs (Supports Figure 4 and 5).  Top 500,000 TF-target predictions were extracted from each network to perform t-SNE clustering using parameter “perplexing=9, permutation=2000”.

\pagebreak
![]($grn/data/14_eval_sum/09.go.heat.pdf){width=100%}

Supplemental Figure 11.  Hierarchical clustering of 98 Gene Ontology (Uniprot.Plants, level 6) terms using fold enrichment in different GRNs (Supports Figure 3-5).  Only GO terms enriched (i.e., P-value < 0.05 and fold change > 1, top 100k edges taken) in at least 3 out of the 44 networks were used for clustering.  Insignificant enrichment (p-value < 0.05) was treated as missing data and shown as white space in the figure.  Fold enrichment is determined as described in Figure 3.  Box areas highlighted in red stand for three sets of GO terms associated with different biological processes: ribosomal assembly and translation processes (A), photosynthetic activity-related processes (B) and the seed oil-body biogenesis term with a kernel- and endosperm- specific function (C).

\pagebreak
![]($grn/data/16_pathways/25.antho.pdf){width=100%}

Supplemental Figure 12.  Different GRNs support different parts of the anthocyanin biosynthesis pathway (Supports Figure 5).

\pagebreak
![]($grn/data/16_pathways/25.dimboa.pdf){width=100%}

Supplemental Figure 13.  Different GRNs support different parts of the DIMBOA pathway (Supports Figure 5).

\pagebreak
![]($grn/data/16_pathways/25.hb26.pdf){width=100%}

Supplemental Figure 14.  Different GRNs support different parts of the chlorophyllide biosynthesis pathway regulated by homeobox-transcription factor 26 (HB26, Zm00001d008612) (Supports Figure 5).

\pagebreak
![]($grn/data/16_pathways/26.pdf){width=100%}

Supplemental Figure 15.  Different coexpression-based GRNs capture distinct parts of classic and CornCyc metabolic pathways (Supports Figure 5).  (A) The teosinte branched1 (tb1)-mediated pathway regulating bud dormancy and growth repression.  (B) The glycolysis pathway.  (C-D) The zealexin biosynthesis pathway (C) regulated by a WRKY-transcription factor 34 (WRKY34, Zm00001d009939) (D).  (E-F) The methylerythritol phosphate pathway (E) regulated by a C2C2-CO-like-transcription factor 11 (COL11, Zm00001d003162) (F).  The letters along the edges of the networks in (B) (D) and (F) indicate significant support from specific GRN as indicated in the key.

\pagebreak
Supplemental Figure 16.  TF-target validation of the combined tissue  network in all six selected natural variation datasets (Supports Figure 6).  Each panel shows the proportion of differentially expressed targets regulated by TFs showing different DE levels between two genotypes in one tissue/treatment condition.  TF-target predictions were binned to 10 groups based on the interaction score in GRN.  Each TF-target pair is classified according to the DE level of the TF (“non_DE”, “DE1-2”, “DE2-4”, “DE4+” or “SPE”) in each network.  The proportion of TF-target pairs with the target also showing DE was then determined for each category.  Within each panel the actual numbers of TF-target pairs falling into the “SPE” category (i.e., purple line) were labelled next to each point.  Dashed line in each panel represents the genome-wide (background) proportion of DE genes in each tissue/treatment setting.

\pagebreak
![]($grn/data/14_eval_sum/30.eQTL.pdf){width=100%}

Supplemental Figure 17. Enrichment of co-regulated targets between previously identified trans-eQTL hotspots and TF-target associations predicted by GRNs (Supports Figure 8).  For each network the top 1 million TF-target predictions were binned to 10 groups based on the interaction score in GRN.  Fold enrichment is determined by the same permutation approach described in Figure 2.

\pagebreak
![]($grn/data/14_eval_sum/31.eQTL.heat.pdf){width=50%}

Supplemental Figure 18. Enrichment of co-regulated targets between previously identified trans-eQTL hotspots and TF-target associations predicted by GRNs (Supports Figure 8).  For each network the top 1 million predicted TF-target associations were binned to 10 bins and only the first bin (top 100k edges) were used to assess enrichment.  15 high quality networks (marked with crosses) were selected to identify the main TF regulators underlying trans-eQTL hotspots (see Methods).

\pagebreak
![]($grn/data/14_eval_sum/32.hs.pdf){width=100%}

Supplemental Figure 19.  Co-localization of TFs predicted by GRNs in this study and trans-eQTL hotspots identified in previous studies that regulate the same set of targets (Supports Figure 8).  Each trans-eQTL hotspot were first tested for significant overlap in targets with any TFs (hypergeometric enrichment test as implemented in the phyper() function in R).  To control for false positives, only TFs identified in at least two (out of 8 high quality networks, see Methods) that show significant co-regulation with at least one trans-eQTL hotspot (p < 0.01) were kept.  Trans-eQTL hotspots identified in previous maize assemblies were lifted over to the AGPv4 assembly coordinates.  Co-localization of a TF and a trans-eQTL is determined if the two coordinates are within 50-Mbp distance.  Color of each dot represents the number of common targets between the predicted TF regulator and the trans-eQTL hotspot.

\pagebreak
Supplemental Table 1. ChIP-Seq and DAP-Seq datasets used in this study.
```{r, message=FALSE, echo=FALSE}
readRDS(file.path(dirw, 'st1.rds'))
```

\pagebreak
Supplemental Table 2. TF knockout mutant RNA-Seq datasets used in this study.
```{r, message=FALSE, echo=FALSE}
readRDS(file.path(dirw, 'st2.rds'))
```

\pagebreak
Supplemental Table 3. Natural variation datasets used for validation in this study.
```{r, message=FALSE, echo=FALSE}
readRDS(file.path(dirw, 'st3.rds'))
```

